File: python/test/t_IntervalMesher_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()

for diamond in [False, True]:
    mesher1D = ot.IntervalMesher([5])
    print("mesher1D=", mesher1D)
    mesh1D = mesher1D.build(ot.Interval(-1.0, 2.0), diamond)
    print("mesh1D=", mesh1D)
    ott.assert_almost_equal(mesh1D.getVolume(), 3.0)

    mesher2D = ot.IntervalMesher([5, 5])
    print("mesher2D=", mesher2D)
    mesh2D = mesher2D.build(ot.Interval([-1.0, -1.0], [2.0, 2.0]), diamond)
    print("mesh2D=", mesh2D)
    ott.assert_almost_equal(mesh2D.getVolume(), 9.0)

    mesher3D = ot.IntervalMesher([5] * 3)
    print("mesher3D=", mesher3D)
    mesh3D = mesher3D.build(ot.Interval(3), diamond)
    print("mesh3D=", mesh3D)
    ott.assert_almost_equal(mesh3D.getVolume(), 1.0)
    volume = mesh3D.computeSimplicesVolume()
    if diamond:
        assert mesh3D.getVerticesNumber() == 791
        nv = 3000
        assert mesh3D.getSimplicesNumber() == nv
        ott.assert_almost_equal(volume, [1.0 / nv] * nv)

    mesher4D = ot.IntervalMesher([2] * 4)
    print("mesher4D=", mesher4D)
    try:
        mesh4D = mesher4D.build(ot.Interval(4), diamond)
        print("mesh4D=", mesh4D)
    except RuntimeError:
        print("notyetimpl")
    ott.assert_almost_equal(mesh4D.getVolume(), 1.0)

    mesher5D = ot.IntervalMesher([2] * 5)
    print("mesher5D=", mesher5D)
    try:
        mesh5D = mesher5D.build(ot.Interval(5), diamond)
        print("mesh5D=", mesh5D)
    except RuntimeError:
        print("notyetimpl")
    ott.assert_almost_equal(mesh5D.getVolume(), 1.0)

File: lib/test/t_IntervalMesher_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class IntervalMesher for standard methods
 *
 *  Copyright 2005-2024 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  try
  {
    // The 1D mesher
    IntervalMesher mesher1D(Indices(1, 5));
    fullprint << "mesher1D=" << mesher1D << std::endl;
    Mesh mesh1D(mesher1D.build(Interval(-1.0, 2.0)));
    fullprint << "mesh1D=" << mesh1D << std::endl;
    IntervalMesher mesher2D(Indices(2, 4));
    fullprint << "mesher2D=" << mesher2D << std::endl;
    Mesh mesh2D(mesher2D.build(Interval(Point(2, -1.0), Point(2, 2.0))));
    fullprint << "mesh2D=" << mesh2D << std::endl;
    IntervalMesher mesher3D(Indices(3, 3));
    fullprint << "mesher3D=" << mesher3D << std::endl;
    Mesh mesh3D(mesher3D.build(Interval(Point(3, -1.0), Point(3, 2.0))));
    fullprint << "mesh3D=" << mesh3D << std::endl;
    IntervalMesher mesher4D(Indices(4, 2));
    fullprint << "mesher4D=" << mesher4D << std::endl;
    Mesh mesh4D(mesher4D.build(Interval(Point(4, -1.0), Point(4, 2.0))));
    fullprint << "mesh4D=" << mesh4D << std::endl;
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }

  return ExitCode::Success;
}
